DownFall/Views/Button.swift

```diff
@@ -49,9 +55,79 @@ class Button: SKShapeNode {
     init(size: CGSize,
          delegate: ButtonDelegate,
          identifier: ButtonIdentifier,
+         image: SKSpriteNode,
          precedence: Precedence,
-         fontSize: CGFloat,
-         fontColor: UIColor,
+         shape: ButtonShape,
+         showSelection: Bool = true,
+         disable: Bool = false) {
+        
+        self.delegate = delegate
+        self.identifier = identifier
+        self.showSelection = showSelection
+        
+        // Call super
+        super.init()
+        
+        let rectangle = CGRect(x: -size.width/2, y: -size.height/2, width: size.width, height: size.height)
+        let shadowRect = CGRect(x: -size.width/2, y: -size.height/2 - dropShadowOffset, width: size.width, height: size.height)
+        let cornerRadius = CGFloat(5.0)
+        let shadowShape: SKShapeNode
+        switch shape {
+        case .rectangle:
+            self.path = CGPath(roundedRect: rectangle, cornerWidth: cornerRadius, cornerHeight: cornerRadius, transform: nil)
+            let shadowPath = CGPath(roundedRect: shadowRect, cornerWidth: cornerRadius, cornerHeight: cornerRadius, transform: nil)
+            shadowShape = SKShapeNode(path: shadowPath)
+            
+            let buttonPath = CGPath(roundedRect: rectangle, cornerWidth: cornerRadius, cornerHeight: cornerRadius, transform: nil)
+            self.buttonView = SKShapeNode(path: buttonPath)
+        case .circle:
+            self.path = CGPath(ellipseIn: rectangle, transform: nil)
+            let shadowPath = CGPath(ellipseIn: shadowRect, transform: nil)
+            shadowShape = SKShapeNode(path: shadowPath)
+            
+            let buttonPath = CGPath(ellipseIn: rectangle, transform: nil)
+            self.buttonView = SKShapeNode(path: buttonPath)
+            
+        }
+        
+        // add the image
+        image.zPosition = 0
+        addChildSafely(image)
+        
+        addChildSafely(buttonView)
+        
+        /// add the drop shadow and keep a reference to it
+        shadowShape.color = .storeBlack
+        shadowShape.zPosition = -1
+        self.dropShadow = shadowShape
+        addChild(shadowShape)
+        
+        // set the name to the identifier
+        name = identifier.rawValue
+        isUserInteractionEnabled = true
+        zPosition = 1000
+        
+        
+        //add touch expanding view
+        addChild(touchTargetExpandingView)
+        
+        //enable/disable
+        self.isDisabled = disable
+        
+        // set points for moving the button slightly
+        unpressedPosition = self.frame.center
+        depressedPosition = self.frame.center.translateVertically(-dropShadowOffset)
+        
+        self.color = .lightBarBlue
```

Comment:
I think this logic can be moved to the commonInit

Reply:
Moved some but a lot of this code is specific to the type of button you are creating.